src/MergeContext.js

```diff
@@ -499,58 +557,90 @@ class MergeContext {
-        // the bot will wait for three checks with 'Jenkins(build test).*' names.
-        const needMatching = checksNumber !== undefined;
-        const requiredChecksNumber = needMatching ? checksNumber : requiredContexts.length;
-
-        // An array of [{context, state}] elements
-        let requiredChecks = [];
-        // filter out non-required checks
-        for (let st of combinedStatus.statuses) {
-            if (requiredContexts.find((el) => { return needMatching ?
-                         st.context.startsWith(el.trim()) : (el.trim() === st.context.trim());}))
-                requiredChecks.push({context: st.context, state: st.state, targetUrl: st.target_url, description: st.description});
-        }
-
-        if (requiredChecks.length < requiredChecksNumber || requiredChecks.find(check => check.state === 'pending'))
-            return 'pending';
-
-        const prevLen = requiredChecks.length;
-        requiredChecks = requiredChecks.filter(check => check.state === 'success');
-        const ret = (prevLen === requiredChecks.length);
-        if (ret && andSupplyRequired) {
-            for (let requiredContext of requiredContexts) {
-                // go further only if the passed check context is not a required one
-                if (requiredChecks.find(el => el.context.trim() === requiredContext.trim()) === undefined) {
-                    // go further only if the passed check context matches the required one
-                    const matched = requiredChecks.find(el => el.context.startsWith(requiredContext.trim()));
-                    // Before a 'staged' commit can be applied, it should have all required checks passed.
-                    // We create a new "required" check, taking other attributes like targetUrl
-                    // from an already succeeded (matching) check (there can be several matching checks,
-                    // e.g., from different Jenkins nodes). After that, there will be two checks
-                    // referencing the same targetUrl.
-                    if (matched && !this._dryRun("required status check creation"))
-                        await GH.createStatus(ref, "success", matched.targetUrl, matched.description, requiredContext);
-                }
-            }
+
+        return requiredContexts;
+    }
+
+    // returns filled StatusResult object
+    async _getPRStatuses() {
+        let requiredContexts = await this._getRequiredContexts();
+        requiredContexts = requiredContexts.filter(el => el.trim() !== Config.approvalContext());
+        const combinedPrStatus = await GH.getStatuses(this._prHeadSha());
+        let statusResult = new StatusResult(requiredContexts.length, "PR");
+        // fill with required status checks
+        for (let st of combinedPrStatus.statuses) {
+            if (requiredContexts.some(el => el.trim() === st.context.trim() && el.trim() !== Config.approvalContext()))
```

Comment:
The "not approvalContext()" condition should be removed both because the "PR Approval" check is not really special and because that check is counted in requiredContexts.length.

Reply:
> This check is 'special' (at least for now) because this is the only check which may be delayed.

I see nothing special here. Virtually all checks take some time to go from pending to final. I suspect you attach some special meaning to the delay of this check because Anubis is responsible for ending the corresponding wait; for me, that responsibility lies outside the StatusResult scope.

None of the two proposed solutions make sense to me. What is wrong with:

* Remove any special handling of the "PR Approval" status check from the StatusResult code. For that class (and its object creators), the check is not special.
* (Continue to?) compute approval and approval-related waiting delays separately, based on the existing Approval class. That computation and related status check updates should happen _before_ StatusResult code gets status checks so that StatusResult gets a fresh "PR Approval" status check.

Reply:
> (Continue to?) compute approval [...]

This is already done. Here is why (IMO) this does not work:

1. _checkMergeConditions() proceeds up to approval checking calculate them. E.g., it is 'slow burner' with a delay.
2. Since the approval is still pending, we update approval status check with fresh 'pending'.
3. Status check method starts (e.g., all other statuses are success, except PR approval, which is pending), so it returns pending.
4. _checkMergeConditions() notices that the aggregate status check is 'pending' and suspends the bot (instead of delaying with the interval calculated at (1)).

We could return 'delayed' after (2) (thus skipping status checking), but this approach is wrong either.

Reply:
> _checkMergeConditions() notices that the aggregate status check is 'pending' and suspends the bot (instead of delaying with the interval calculated at (1)).

How about changing this logic to suspend the bot only if the earlier approval check did not result in a delayed approval? Overall principle: We should never suspend (i.e., trigger an infinite wait) if there is change expected after a finite wait.